File: ruby/src/sorting/quick_sort.rb
require_relative '../common/array_utils'

module Sorting
  class << self

    def quick_sort(array, opts = {}, &comparator)
      start_index = opts[:start_index] || 0
      end_index = opts[:end_index] || array.length - 1

      return array if start_index >= end_index

      pivot_index = partition(array,
                              :start_index => start_index,
                              :end_index => end_index,
                              &comparator)
      array = quick_sort(array,
                         :start_index => start_index,
                         :end_index => pivot_index - 1,
                         &comparator)
      array = quick_sort(array,
                         :start_index => pivot_index + 1,
                         :end_index => end_index,
                         &comparator)
      array
    end

    def random_pivot(start_index, end_index)
      length = end_index - start_index + 1
      Random.rand(length) + start_index
    end

    def partition(array, opts = {}, &comparator)
      start_index = opts[:start_index] || 0
      end_index = opts[:end_index] || array.length - 1
      pivot_index = opts[:pivot_index] || random_pivot(start_index, end_index)

      # Put the pivot in the end.
      ArrayUtils.swap(array, pivot_index, end_index)

      swap_index = start_index - 1

      # When the loop below terminates, we will have
      # all elements <= pivot in the range (0..swapindex).
      (start_index..end_index - 1).each do |index|
        comparison = yield(array[index], array[end_index])
        if comparison <= 0
          swap_index += 1
          ArrayUtils.swap(array, swap_index, index)
        end
      end

      # Put the pivot at the position where it deserves to be in.
      ArrayUtils.swap(array, swap_index + 1, end_index)
      swap_index + 1
    end
  end
end

File: java/src/sorting/QuickSort.java
package sorting;

import java.util.Comparator;
import java.util.Random;

public class QuickSort {
	public static <T> T[] quickSort(T[] array, Comparator<T> comparator) {
		return quickSort(array, 0, array.length - 1, comparator);
	}

	private static <T> T[] quickSort(T[] array, int startIndex, int endIndex,
			Comparator<T> comparator) {
		if (startIndex >= endIndex) {
			return array;
		}

		int pivotIndex = partition(array, startIndex, endIndex, comparator, -1);
		array = quickSort(array, startIndex, pivotIndex - 1, comparator);
		array = quickSort(array, pivotIndex + 1, endIndex, comparator);
		return array;
	}

	private static int getRandomPivotIndex(int startIndex, int endIndex) {
		int length = endIndex - startIndex + 1;
		return new Random().nextInt(length) + startIndex;
	}

	public static <T> int partition(T[] array, int startIndex, int endIndex,
			Comparator<T> comparator, int pivotIndex) {
		if (pivotIndex == -1) {
			pivotIndex = getRandomPivotIndex(startIndex, endIndex);
		}

		common.ArrayUtils.swap(array, pivotIndex, endIndex);
		int swapIndex = startIndex - 1;
		for (int index = startIndex; index <= endIndex - 1; index++) {
			if (comparator.compare(array[index], array[endIndex]) <= 0) {
				swapIndex++;
				common.ArrayUtils.swap(array, swapIndex, index);
			}
		}

		common.ArrayUtils.swap(array, swapIndex + 1, endIndex);
		return swapIndex + 1;
	}
}
